desktop/createDownloadQueue.ts

```diff
@@ -0,0 +1,101 @@
+import type {BrowserWindow} from 'electron';
+import {app} from 'electron';
+import * as path from 'path';
+import createQueue from '@libs/Queue/Queue';
+import ELECTRON_EVENTS from './ELECTRON_EVENTS';
+import type {Options} from './electronDownloadManagerType';
+
+type DownloadItem = {
+    // The window where the download will be initiated
+    win: BrowserWindow;
+
+    // The URL of the file to be downloaded
+    url: string;
+
+    // The options for the download, such as save path, file name, etc.
+    options: Options;
+};
+
+/**
+ * Returns the filename with extension based on the given name and MIME type.
+ * @param name - The name of the file.
+ * @param mime - The MIME type of the file.
+ * @returns The filename with extension.
+ */
+const getFilenameFromMime = (name: string, mime: string): string => {
+    const extensions = mime.split('/').pop();
+    return `${name}.${extensions}`;
+};
+
+const createDownloadQueue = () => {
+    const downloadItemProcessor = (item: DownloadItem): Promise<void> =>
+        new Promise((resolve, reject) => {
+            item.win.webContents.downloadURL(item.url);
+
```

Comment:
Should we set a timeout here as well and reject the promise if the listener does not get called within that timeout period?

Reply:
@allroundexperts What is the expected behavior if the loading spinner timeout is reached? Should it stop downloading or continue downloading even after the spinner has stopped? Rejecting here will process the next item, resulting in some files not being downloaded due to the original issue in this PR.

Reply:
I'm not sure if I understand your question. In the processor, if the listener is not called after a certain amount of time, then that really means something is wrong. Isn't that the case?

Reply:
I think we are not on the same page here. My understanding is that `item.win.webContents.session.on('will-download', listener);` would be called immediately on download start. You're suggesting that it would be called once the download completes?

Reply:
@wildan-m You should remove the timeout if listener gets called. That was the whole point of adding it.

Reply:
```
   const downloadItemProcessor = (item: DownloadItem): Promise<void> =>
        new Promise((resolve, reject) => {
            item.win.webContents.downloadURL(item.url);

            // Add a timeout to resolve the promise after 5 seconds
            const timeout = setTimeout(() => {
                item.win.webContents.session.removeListener('will-download', listener);
                resolve();
            }, 5000);

            const listener = (event: Electron.Event, electronDownloadItem: Electron.DownloadItem) => {
               clearTimeout(timeout)
                .................

                item.win.webContents.send(ELECTRON_EVENTS.DOWNLOAD_STARTED, {url: item.url});
            };

            item.win.webContents.session.on('will-download', listener);

        });
```

Reply:
@allroundexperts clearing the timeout will not resolve the issue I [mentioned earlier in the video](https://github.com/Expensify/App/pull/37131#discussion_r1595218712).

Reply:
Have you tried it? Why doesn't it solve the issue? Does the `listener` not get called even after `5s`?

Reply:
@allroundexperts I apologize for the incorrect assumption. I thought that will-download would not be called when the saveAs dialog opens. I have made the updates, please review them.